pkg/controllers/raycluster_webhook_test.go

```diff
@@ -0,0 +1,615 @@
+									},
+									VolumeMounts: certVolumeMounts(),
+								},
+							},
+						},
+					},
+					RayStartParams: map[string]string{},
+				},
+			},
+		},
+	}
+
+	invalidRayCluster := &rayv1.RayCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-invalid-raycluster",
+			Namespace: namespace,
+		},
+		Spec: rayv1.RayClusterSpec{
+			HeadGroupSpec: rayv1.HeadGroupSpec{
+				Template: corev1.PodTemplateSpec{
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{},
+					},
+				},
+				RayStartParams: map[string]string{},
+			},
+			WorkerGroupSpecs: []rayv1.WorkerGroupSpec{
+				{
+					GroupName: "worker-group-1",
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Name: "worker-container-1",
+								},
+							},
+							InitContainers: []corev1.Container{
+								{
+									Name:  "create-cert",
+									Image: "quay.io/project-codeflare/ray:latest-py39-cu118",
+									Command: []string{
+										"sh",
+										"-c",
+										`cd /home/ray/workspace/tls && openssl req -nodes -newkey rsa:2048 -keyout server.key -out server.csr -subj '/CN=ray-head' && printf "authorityKeyIdentifier=keyid,issuer\nbasicConstraints=CA:FALSE\nsubjectAltName = @alt_names\n[alt_names]\nDNS.1 = 127.0.0.1\nDNS.2 = localhost\nDNS.3 = ${FQ_RAY_IP}\nDNS.4 = $(awk 'END{print $1}' /etc/hosts)">./domain.ext && cp /home/ray/workspace/ca/* . && openssl x509 -req -CA ca.crt -CAkey ca.key -in server.csr -out server.crt -days 365 -CAcreateserial -extfile domain.ext`,
+									},
+									VolumeMounts: certVolumeMounts(),
+								},
+							},
+						},
+					},
+					RayStartParams: map[string]string{},
+				},
+				{
+					GroupName: "worker-group-2",
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Name: "worker-container-2",
+								},
+							},
+						},
+					},
+					RayStartParams: map[string]string{},
+				},
+			},
+		},
+	}
+
+	// Create the RayClusters
+	test.Client().Ray().RayV1().RayClusters(namespace).Create(test.Ctx(), validRayCluster, metav1.CreateOptions{})
+	test.Client().Ray().RayV1().RayClusters(namespace).Create(test.Ctx(), invalidRayCluster, metav1.CreateOptions{})
+
+	// Positive test case (expect no errors)
+	validWorkerInitContainer := validateWorkerInitContainer(validRayCluster)
+	test.Expect(validWorkerInitContainer).To(gomega.BeEmpty())
+
+	// Negative test case (expect errors for missing Init Container)
+	invalidWorkerInitContainer := validateWorkerInitContainer(invalidRayCluster)
+	test.Expect(invalidWorkerInitContainer).NotTo(gomega.BeEmpty())
+	test.Expect(invalidWorkerInitContainer[0].Detail).To(gomega.ContainSubstring("create-cert Init Container is immutable"))
```

Comment:
One general comment - can you split test cases, so every test case is validated separately?
i.e. here you validate both valid and invalid case, so in case positive test case fails then negative test case is not validated.

Reply:
Ohh that's right! Thanks - will split test cases.